cli/src/cmd/forge/geiger/visitor.rs

```diff
@@ -0,0 +1,336 @@
+use crate::cmd::forge::geiger::find::CheatcodeCounter;
+use forge_fmt::{Visitable, Visitor};
+use solang_parser::pt::{
+    ContractDefinition, Expression, FunctionDefinition, IdentifierPath, Loc, Parameter, SourceUnit,
+    Statement, TypeDefinition, VariableDeclaration, VariableDefinition,
+};
+use std::convert::Infallible;
+
+/// a [`forge_fmt::Visitor` that scans for invocations of cheatcodes
+#[derive(Default)]
+pub struct CheatcodeVisitor {
+    pub cheatcodes: CheatcodeCounter,
+}
+
+impl Visitor for CheatcodeVisitor {
+    type Error = Infallible;
+
+    fn visit_source_unit(&mut self, source_unit: &mut SourceUnit) -> Result<(), Self::Error> {
+        source_unit.0.visit(self)
+    }
+
+    fn visit_contract(&mut self, contract: &mut ContractDefinition) -> Result<(), Self::Error> {
+        contract.base.visit(self)?;
+        contract.parts.visit(self)
+    }
+
+    fn visit_block(
+        &mut self,
+        _loc: Loc,
+        _unchecked: bool,
+        statements: &mut Vec<Statement>,
+    ) -> Result<(), Self::Error> {
+        statements.visit(self)
+    }
+
+    fn visit_expr(&mut self, _loc: Loc, expr: &mut Expression) -> Result<(), Self::Error> {
+        match expr {
+            Expression::PostIncrement(_, expr) => {
+                expr.visit(self)?;
+            }
+            Expression::PostDecrement(_, expr) => {
+                expr.visit(self)?;
+            }
+            Expression::New(_, expr) => {
+                expr.visit(self)?;
+            }
+            Expression::ArraySubscript(_, expr1, expr2) => {
+                expr1.visit(self)?;
+                expr2.visit(self)?;
+            }
+            Expression::ArraySlice(_, expr1, expr2, expr3) => {
+                expr1.visit(self)?;
+                expr2.visit(self)?;
+                expr3.visit(self)?;
+            }
+            Expression::Parenthesis(_, expr) => {
+                expr.visit(self)?;
+            }
+            Expression::MemberAccess(_, expr, _) => {
+                expr.visit(self)?;
+            }
+            Expression::FunctionCall(loc, lhs, rhs) => {
+                // all cheatcodes are accessd via <vm>.cheatcode
```

Comment:
Can this assumption be broken using:
```solidity
address(HEVM_ADDRESS).call(malicious_data);
assembly {
    let success = call(args)
}
```

Reply:
Probably. I don't know that we can reliably/fully foolproof the software?

@mattsse sorry I forgot - did we confirm that with `cast run` cheatcodes are disabled?

Reply:
Well given you can `{CALLDATA,RETURNDATA,S}LOAD` and then `CALL`, there probably isn't a way to guarantee all calls are safe at compile time?

Reply:
> [just concerned that this my lull users into a false sense of security]

Right. In that case should we not ship this at all? Or should we make it very explicit that this is just a _helper_ tool and not guaranteed to find everything?

Reply:
perhaps we disable the subcommand until we decide?

Reply:
> Probably. I don't know that we can reliably/fully foolproof the software?
> 
> @mattsse sorry I forgot - did we confirm that with `cast run` cheatcodes are disabled?

yep, no cheatcode inspector for `cast run`

Reply:
I would hide the subcommand. Something like this feels powerful as a tool but the interface should make it clear to the user that it's a tool for scanning solidity function calls, not executing EVM code execution (which i dont think is fundamentally possible).